apm-lambda-extension/extension/http_server_test.go

```diff
@@ -137,3 +143,116 @@ func Test_getDecompressedBytesFromRequestEmptyBody(t *testing.T) {
 		t.Fail()
 	}
 }
+
+/**
+ * Starts a mock HTTP server
+ *
+ * Exposes a single / endpoint.  This endpoint returns
+ * a JSON string with two properties
+ *
+ * data: the value passed via `response`, as JSON
+ * seen_headers: the headers seen by the request to /
+ *
+ * The intended use is to start the extension's HTTP server
+ * and confirm that it proxies info requests to this server
+ */
+func startMockApmServer(response map[string]string, port string) {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		seenHeaders := make(map[string]string)
+		for name, values := range r.Header {
+			for _, value := range values {
+				seenHeaders[name] = value
+			}
+		}
+
+		data := map[string]map[string]string{
+			"data":         response,
+			"seen_headers": seenHeaders,
+		}
+		responseString, _ := json.Marshal(data)
+		w.Write([]byte(responseString))
+	})
+	log.Fatal(http.ListenAndServe(":"+port, nil))
+}
+
+func getUrl(url string, headers map[string]string, t *testing.T) string {
+	client := &http.Client{}
+	req, _ := http.NewRequest("GET", url, nil)
+	for name, value := range headers {
+		req.Header.Add(name, value)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		t.Logf("Error fetching %s, [%v]", url, err)
+		return "{}"
+	}
+
+	defer resp.Body.Close()
+	body, _ := ioutil.ReadAll(resp.Body)
+	return string(body)
+}
+
+func startExtension(config *extensionConfig) {
+	dataChannel := make(chan []byte, 100)
+	NewHttpServer(dataChannel, config)
+}
+
+func getRandomNetworkPort(notwanted []int) int {
+	rand.Seed(time.Now().UnixNano())
+	min := 50000
+	max := 65534
+	var value = rand.Intn(max-min+1) + min
+	// if we've got a value that's not wanted,
+	// recall outself recursivly
+	for _, v := range notwanted {
+		if value == v {
+			return getRandomNetworkPort(notwanted)
+		}
+	}
+	return value
+}
+
+func TestProxy(t *testing.T) {
+	var iApmServerPort = getRandomNetworkPort([]int{})
+	apmServerPort := fmt.Sprint(iApmServerPort)
+
+	go startMockApmServer(
+		map[string]string{"foo": "bar"},
+		apmServerPort,
+	)
+
+	var extensionPort = fmt.Sprint(getRandomNetworkPort([]int{iApmServerPort}))
+
+	go startExtension(&extensionConfig{
+		apmServerUrl:               "http://localhost:" + apmServerPort,
+		apmServerSecretToken:       "foo",
+		apmServerApiKey:            "bar",
+		dataReceiverServerPort:     ":" + extensionPort,
+		dataReceiverTimeoutSeconds: 15,
+	})
+
+	// sleep seem neccesary in our test env
+	// to give the above go-routines-servers
+	// time to complete their startup
+	time.Sleep(1 * time.Second)
```

Comment:
I've seen in the apm server repo that they use `WaitGroup`s for this. I think it would be a preferable approach over `Sleep`ing.